### PR TITLE
MeshIO using new Properties and PropertyVector.

### PR DESCRIPTION
**Pull request by TomFischer**
Reading material ids of elements from the old ogs mesh file format (*.msh) into the new data structures.

### CODE REVIEW

File: FileIO/Legacy/MeshIO.cpp [hunk 1]
```diff
@@ -119,6 +123,17 @@ MeshLib::Mesh* MeshIO::loadMeshFromFile(const std::string& file_name)
 		MeshLib::Mesh* mesh (new MeshLib::Mesh(BaseLib::extractBaseNameWithoutExtension(
 		                                               file_name), nodes, elements));
 
+		boost::optional<MeshLib::PropertyVector<int> &> opt_material_ids(
+			mesh->getProperties().createNewPropertyVector<int>(
+				"MaterialIDs", MeshLib::MeshItemType::Cell, 1)
+		);
+		if (!opt_material_ids) {
+			WARN("No material ids read.");
```
Comment (rinkk) [inline FileIO/Legacy/MeshIO.cpp]: is this possible? and if so, is it still a valid mesh file?

File: FileIO/Legacy/MeshIO.cpp [hunk 2]
```diff
@@ -119,6 +123,17 @@ MeshLib::Mesh* MeshIO::loadMeshFromFile(const std::string& file_name)
...
+		} else {
+			MeshLib::PropertyVector<int> & material_ids(opt_material_ids.get());
+			material_ids.resize(materials.size());
+			std::copy(materials.cbegin(), materials.cend(), material_ids.begin());
```
Comment (rinkk) [inline FileIO/Legacy/MeshIO.cpp]: stupid question: since ```material_ids``` is empty at this point, wouldn't it be faster to append the mat-vector, e.g.
```material_ids.insert(material_ids.end(), materials.begin(), materials.end());``` ?

File: FileIO/Legacy/MeshIO.cpp [hunk 3]
```diff
@@ -133,18 +148,22 @@ MeshLib::Mesh* MeshIO::loadMeshFromFile(const std::string& file_name)
 	}
 }
 
-MeshLib::Element* MeshIO::readElement(const std::string& line,
-                                      const std::vector<MeshLib::Node*> &nodes)
+std::size_t MeshIO::readMaterialID(std::istream & in) const
+{
+	unsigned index, patch_index;
```
Comment (rinkk) [inline FileIO/Legacy/MeshIO.cpp]: why is the variable called ```patch_index``` and not ```material``` or so?

File: FileIO/Legacy/MeshIO.cpp [hunk 4]
```diff
@@ -133,18 +148,22 @@ MeshLib::Mesh* MeshIO::loadMeshFromFile(const std::string& file_name)
...
 	MeshElemType elem_type (MeshElemType::INVALID);
-	unsigned index, patch_index;
-	ss >> index >> patch_index;
 
 	do {
-		ss >> elem_type_str;
-		if (ss.fail())
+		in >> elem_type_str;
+		if (in.fail())
 			return NULL;
```
Comment (rinkk) [inline FileIO/Legacy/MeshIO.cpp]: nullptr

File: FileIO/Legacy/MeshIO.cpp [hunk 5]
```diff
@@ -133,18 +148,22 @@ MeshLib::Mesh* MeshIO::loadMeshFromFile(const std::string& file_name)
...
 	std::string elem_type_str("");
 	MeshElemType elem_type (MeshElemType::INVALID);
-	unsigned index, patch_index;
-	ss >> index >> patch_index;
 
 	do {
-		ss >> elem_type_str;
-		if (ss.fail())
+		in >> elem_type_str;
+		if (in.fail())
```
Comment (rinkk) [inline FileIO/Legacy/MeshIO.cpp]: why do you need to test here for a ```fail``` but not in the ```readMaterialID()```-method?

File: FileIO/Legacy/MeshIO.cpp [hunk 6]
```diff
@@ -133,18 +148,22 @@ MeshLib::Mesh* MeshIO::loadMeshFromFile(const std::string& file_name)
...
 			return NULL;
 		elem_type = String2MeshElemType(elem_type_str);
 	} while (elem_type == MeshElemType::INVALID);
```
Comment (rinkk) [inline FileIO/Legacy/MeshIO.cpp]: sorry for the many questions but why do we need a ```do ... while``` here? (i see it is old code and presumably there is a reason for this but i cannot think of any ...)

File: FileIO/Legacy/MeshIO.cpp [hunk 7]
```diff
@@ -156,66 +175,66 @@ MeshLib::Element* MeshIO::readElement(const std::string& line,
 	{
 	case MeshElemType::LINE: {
 		for (int i = 0; i < 2; ++i)
-			ss >> idx[i];
+			in >> idx[i];
```
Comment (rinkk) [inline FileIO/Legacy/MeshIO.cpp]: for this and all the other element types: what happens if you run out of values to read? (e.g. you need to read two node ids for a line but what happens if the stream ends after the first value?)

File: FileIO/Legacy/MeshIO.h [hunk 8]
```diff
@@ -57,7 +57,8 @@ class MeshIO : public Writer
 
 private:
 	void writeElements(std::vector<MeshLib::Element*> const& ele_vec, std::ostream &out) const;
-	MeshLib::Element* readElement(const std::string& line, const std::vector<MeshLib::Node*> &nodes);
+	std::size_t readMaterialID(std::istream & in) const;
+	MeshLib::Element* readElement(std::istream& line, const std::vector<MeshLib::Node*> &nodes) const;
```
Comment (rinkk) [inline FileIO/Legacy/MeshIO.h]: documentation for both methods is missing ... the name is pretty obvious so i don't know if something is necessary or not ...

**Comment by rinkk** [comment]
Two general things:
1. is it necessary to adjust the write method, too?
2. to have a robust read method, i think it would be best to do something like this:
  * read line
  * subdivide line into list of values
  * read 3rd value of line (element type id) and evaluate if number of read values is equal to numbers that need to be read (e.g. 7 values for a quad, etc.)
  * build elements based on these value.

this would require a lot of changes to the current read method and i really don't know if it is worth the effort for a legacy interface. i am just suggesting that the current implementation might crash in a number of cases. but it's also okay if we decide not to care about that ....

**Comment by rinkk** [comment]
Another thing: we currently don't have tests for this. I am not sure if we need some (legacy interface and all that). if you feel like it you could write some ...

File: FileIO/Legacy/MeshIO.cpp (hunk 1 above)
Comment (TomFischer) [inline FileIO/Legacy/MeshIO.cpp]: It should not be possible. But ...

File: FileIO/Legacy/MeshIO.cpp (hunk 3 above)
Comment (TomFischer) [inline FileIO/Legacy/MeshIO.cpp]: Afaik GMSH uses patch_index as material id. But you are right: for readability the variable should be renamed.

**Comment by TomFischer** [comment]
In my opinion this is a legacy file format. I do not want to put much more effort than using the new data structures for properties, i.e. I do not want write test nor I want to improve the robustness.

With the new commits I hope I addressed all your other comments.

File: FileIO/Legacy/MeshIO.cpp [hunk 9]
```diff
@@ -103,7 +105,9 @@ MeshLib::Mesh* MeshIO::loadMeshFromFile(const std::string& file_name)
 				for (unsigned i = 0; i < nElements; ++i)
 				{
 					getline(in, line_string);
-					elements.push_back(readElement(line_string, nodes));
+					std::stringstream ss(line_string);
+					materials.push_back(readMaterialID(ss));
+					elements.push_back(readElement(ss, nodes));
```
Comment (norihiro-w) [inline FileIO/Legacy/MeshIO.cpp]: no need to stop reading if `readElement()` returns nullptr?

Comment (TomFischer) [inline FileIO/Legacy/MeshIO.cpp]: I did not change the already implemented behaviour. What do you think? Opinions of other developers are also wellcome. If we decide to change the behaviour should we do this in a separate PR?

Comment (rinkk) [inline FileIO/Legacy/MeshIO.cpp]: as i mentioned in my comments there are multiple ways to break this read-method. however, in my opinion we an leave it as it is for the most part as it is "only" a legacy interface and ogs5-read methods are similarly buggy (i.e. it is unlikely that someone comes up with a msh-file that will read in ogs5 but not in ogs6).

that said, in this particular case i would also *not* insert a nullptr but instead either abort reading the mesh or simply skip the element completely. in both cases we can guarantee consistent behaviour afterwards, whereas a nullptr within the element-vector will result in a crash later on...

**Comment by bilke** [comment]
:+1:

File: FileIO/Legacy/MeshIO.cpp (hunk 9 above)
Comment (TomFischer) [inline FileIO/Legacy/MeshIO.cpp]: Changed as requested.
